minitask2/benh/robot_control.py

```diff
@@ -0,0 +1,289 @@
+        '''
+        self.move(0, 0)
+
+    @abstractmethod
+    def shouldrun(self, scandata):
+        pass
+
+    @abstractmethod
+    def run(self, scandata):
+        pass
+
+    @abstractmethod
+    def clear_behaviour(self):
+        pass
+
+
+class ObstacleAvoidance(RobotBehaviour):
+
+    TURN_SPEED = 0.3
+
+    def __init__(self, vel_pub):
+        super(ObstacleAvoidance, self).__init__(vel_pub)  # Python2 is bad
+        self.name = "Obstacle Avoidance"
+
+    def shouldrun(self, scandata):
+        '''
+        Returns true if there is an object closer than 0.5m in front of the robot.
+        '''
+        return super(ObstacleAvoidance, self).sample_scan(scandata,
+                                                          0,
+                                                          window_size=60) <= 0.5
+
+    def run(self, scandata):
+        super(ObstacleAvoidance, self).turn(ObstacleAvoidance.TURN_SPEED)
+
+    def clear_behaviour(self):
+        pass
+
+
+class WallFollowing(RobotBehaviour):
+
+    START_THRESHOLD = 1.0
+    IDEAL_DIST = 1.0
+    END_THRESHOLD = 1.5
+    SAMPLE_ANGLE = 10
+    WINDOW = 8  # Degrees
+
+    MOVE_SPEED = 0.1
+    TURN_GAIN = 14
+
+    def __init__(self, vel_pub):
+        super(WallFollowing, self).__init__(vel_pub)  # Python2 is bad
+        self.name = "Wall Following"
+        self.started = False
+
+    def shouldrun(self, scandata):
+        '''
+        Takes 3 samples, if they are all under the start threshold, then start
+        '''
+        right = super(WallFollowing,
+                      self).sample_scan(scandata,
+                                        270,
+                                        window_size=WallFollowing.WINDOW)
+        top = super(WallFollowing,
+                    self).sample_scan(scandata,
+                                      270 - WallFollowing.SAMPLE_ANGLE,
+                                      window_size=WallFollowing.WINDOW)
+        if self.started:
+            return all((right <= WallFollowing.END_THRESHOLD,
+                        top <= WallFollowing.END_THRESHOLD))
+        else:
+            return all((right <= WallFollowing.START_THRESHOLD,
+                        top <= WallFollowing.START_THRESHOLD))
+
+    def run(self, scandata):
+        self.started = True
+
+        right = super(WallFollowing,
+                      self).sample_scan(scandata,
+                                        270,
+                                        window_size=WallFollowing.WINDOW)
+        top = super(WallFollowing,
+                    self).sample_scan(scandata,
+                                      270 - WallFollowing.SAMPLE_ANGLE,
+                                      window_size=WallFollowing.WINDOW)
+
+        # This is what the top angle should be if the wall is straight and we
+        # are parallel to it
+        predicted_top = right / cos(radians(WallFollowing.SAMPLE_ANGLE))
+        err = top - predicted_top
```

Comment:
Correct me if I'm wrong.
`right` is the `min` distance to object in scan window 266-274 and
`top` is the `min` distance to object in scan window 256-264 then
`top` window is actually lower than `right` window and it should be `bottom`?

Reply:
Good catch!

I think its working because the sign (positive/negative) on the turn must be wrong as well, I'll fix this.